core/Cargo.toml

```diff
@@ -380,6 +380,7 @@ probe = { version = "0.5.1", optional = true }
 getrandom = { version = "0.2", features = ["js"] }
 
 [dev-dependencies]
+codspeed-criterion-compat = { version = "2.6.0", features = ["async", "async_tokio"] }
```

Comment:
I'm surprised that we have to rely on the `codspeed` crate in this manner. Can we implement this workflow without touching `core` code?

Reply:
I didn't find the way in codspeed document that could not rely on `codspeed`'s crate to implement that. Maybe we could use a compile feature gate to enable it only on the ci? /cc @adriencaccia Do you have any suggestion about that?

Reply:
> Maybe we could use a compile feature gate to enable it only on the ci?

Hi, this doesn't improve the situation.

I expect a single workflow to handle all these tasks so that we can remove, switch, or disable it as needed without modifying the code. From my perspective, this is crucial in helping to reduce our maintenance burden.

Reply:
Services come and go, but our code always remains here. We should focus on our code rather than the service itself.

Reply:
I apologize for not reviewing this PR more thoroughly earlier.

Reply:
I'm fine. I originally saw this PR as an opportunity for a quick try, but we've put a lot of effort into the seeking permit process. BTW, Arrow side seems to face similar issues.😢

Reply:
> the seeking permit process.

CodeSpeed requires the installation of the GitHub app, which significantly prolongs the process. The team members at ASF INFRA need to thoroughly review this app from various perspectives. I'm also tired of the delays, but considering that ASF manages 2.7k repositories and 5.4k people, installing an app could impact all of them. I can understand every concern they might have.

---

Anyways, we now have this app approved. The only obstacle is the code changes. We hope to find a way to use CodeSpeed without modifying the code.

Reply:
@dqhl76 @Xuanwo if using what I outlined in https://github.com/apache/opendal/pull/4993#discussion_r1808981247, the changes to the source code (the `.rs` files) can be removed from this PR, and the workflow will work.

So the only modified code would be the `dev-dependencies` section of the `core/Cargo.toml` and the corresponding changes in `core/Cargo.lock`.

At this moment, this is the minimum footprint the codspeed integration can have on the source code of a repository.
For example, [this commit](https://github.com/CodSpeedHQ/pingora/commit/37f83517885ec59ec942ab3b55bc65ed393fb1e8#diff-ece5f678dd051959c8f661189e0ace2a47c82bb119ad47c2882c3399ef35e8e5) shows the minimal changes for the installation.

Reply:
Hi @adriencaccia, can we implement codspeed without making any changes to the `core`? I'm open to discuss with the upstream `criterion` team to devise a plan for integrating codspeed.

Reply:
Hey @Xuanwo, unfortunately, we need to replace the `criterion` with `codspeed-criterion-compat` in order to make CodSpeed work, and the changes in `codspeed-criterion-compat` are way too specific to get integrated into the upstream `criterion`.

If you do not want to touch at the code of `core`, another solution would be to create a new crate named `benchmark` for example, and have all benchmarks defined there, without impacting the source code `core`.

Reply:
Thank you @adriencaccia for your advice!